Support expires_in in ActiveSupport::Cache::MemCacheStore#increment and #decrement

aeroastro opened the pull request:
### Summary

Previously, `ActiveSupport::Cache::MemCacheStore#increment` and
`ActiveSupport::Cache::MemCacheStore#decrement` call `Dalli::Client#incr`
and `Dalli::Client#decr` as follows.

```ruby
@data.incr(normalize_key(name, options), amount)
@data.decr(normalize_key(name, options), amount)
```

However, considering the method arguments of `Dalli::Client`,
`options[:expires_in]` and default value for a not stored value
are not passed correctly.

* `Dalli::Client#incr(key, amt = 1, ttl = nil, default = nil)`
* `Dalli::Client#decr(key, amt = 1, ttl = nil, default = nil)`

c.f. https://github.com/petergoldstein/dalli/blob/f41bb9cf078246bd58a249cf09b1f56044e4662c/lib/dalli/client.rb#L169-L202

As a result, initial values for not stored values are not correctly set
despite the documentation saying `Calling it on a value not stored
with :raw will initialize that value to zero.`

c.f.
https://github.com/aeroastro/rails/blob/9befc197f926272abbba5a1ca1323ce4f15ebd10/activesupport/lib/active_support/cache/mem_cache_store.rb#L117-L120

Fortunately, since `ttl` is only applicable when the value is already set,
we have not encountered any bugs regarding `expires_in`.

By this patch, `ActiveSupport::Cache::MemCacheStore` will correctly handle
initialization of counters. Also, test on increment/decrement behavior will be
fixed.

### Other Information

P.S. Regarding this initialization, I have posted feature-request on the Google Group to initialize the not stored value to amount as if it is initialized to 0 **prior to the operation**.

sgrif commented:
Unless we can support this for all adapters, I don't think this is a feature worth adding.

aeroastro commented:
@sgrif 

Thank you for reviewing this. 🐱 

I have also been wondering why only `MemCacheStore` supports initialization. It would be much better that all the adapters support initialization.

I will try to address that issue in another PR!

However, on this PR, I just would like to fix the inconsistency between document and implementation before moving toward feature implementation.
https://github.com/aeroastro/rails/blob/9befc197f926272abbba5a1ca1323ce4f15ebd10/activesupport/lib/active_support/cache/mem_cache_store.rb#L117-L120

Which do you think would be better, to remove the document or to fix the implementation?
Any other ideas are welcome.

aeroastro commented:
@sgrif

Hello.

It's been a while since I submitted this PR, and I'm wondering if there is anything I could do on my end. I totally understand you are busy and require more time to consider, it is much appreciated if you could give me a response.

I think there are 3 ways to address this issue.

1. Fix bug here and then introduce new features to the other store classes.
2. Fix document here and then introduce new features to the other store classes.
3. Fix bug of `MemCacheStore` with introducing new features to the other stores simultaneously.

aeroastro commented:
@jeremy

As you are implementing the Redis store with initialization to `amount` (not `nil` nor `0`) at #31134, I also would like your advice on this bug fix on Memcache Store.


Hopefully, I would like to do the followings.

1. Fix the bug on Memcache Store by this PR.
2. Make another PR to set default to `amount` not `0` like Redis store.
https://groups.google.com/forum/#!topic/rubyonrails-core/xThQp862ymI

If required, I will do both bug fix and feature implementation here.

jeremy commented:
Thanks @aeroastro. Let's split into separate PRs as you suggest: 1. this PR, support expires_in/TTL on counters (needs a test case too?) and 2. a new PR, set `default` to `amount`.

aeroastro commented:
@jeremy 

Thank you for your feedback.
I have limited the changes to supporting ttl/expires_in in MemCacheStore here.

If this is O.K., I will rebase and fixup the changes to 1 commit to be merged.

Since there is no test checking middleware's ttl, probably because it consumes actual wall clock time, I have not added test regarding this ttl/expires_in.

jeremy commented:
> If this is O.K., I will rebase and fixup the changes to 1 commit to be merged.

👍

> Since there is no test checking middleware's ttl, probably because it consumes actual wall clock time, I have not added test regarding this ttl/expires_in.

Perhaps use `assert_called` to verify that `expires_in` is passed to the `incr` call?

aeroastro committed:
Support expires_in in ActiveSupport::Cache::MemCacheStore#increment and #decrement

aeroastro commented:
Thank you for the feedback.
I added some tests on `expires_in` at ff98ccf using `assert_called`.
After that I have fixed them up into 1 commit to be merged.

aeroastro commented:
@jeremy 
Thank you for your thumbs up emoji!
But, it seems like this PR is mistakenly closed instead of being merged. 😿  Or the commit will be cherry-picked in some other places?

matthewd commented:
@aeroastro Jeremy did a squash merge: b22ee64b5b30c6d5039c292235e10b24b1057f6d, linked from the "closed" message, is on master.

aeroastro commented:
@matthewd @jeremy 
Oh, I see. There it is!
Thank you again!  🐱
